yakut/param/formatter.py

```diff
@@ -66,21 +69,99 @@ def _make_json_formatter() -> Formatter:
     return lambda data: typing.cast(str, json.dumps(data, ensure_ascii=False, separators=(",", ":")))
 
 
+def _insert_format_specifier(items: typing.List[typing.Tuple[str, typing.Any]], key, instance, is_start: bool = True):
+    is_list = isinstance(instance, Collection) and not isinstance(instance, str)
+    is_dictionary = isinstance(instance, Mapping)
+    if is_start:
+        if is_dictionary:
+            items.append((key + "{", "{"))
+        elif is_list:
+            items.append((key + "[", "["))
+    else:
+        if is_dictionary:
+            items.append((key + "}", "}"))
+        elif is_list:
+            items.append((key + "]", "]"))
+
+
+def flatten_start(
+        d: typing.Union[typing.Dict[typing.Any, typing.Any], typing.Collection[typing.Any]],
+        parent_key: str = "",
+        sep: str = ".",
+        do_put_format_specifiers: bool = False,
```

Comment:
This is fragile. I strongly suggest that you replace this imperative-style solution with a functional-style alternative where instead of the _state_ you specify _behavior_ in the form of a function that is invoked whenever the flattener descends/ascends the data tree. Also, that function does not necessarily have to accept the container to modify, since that [expands its context unnecessarily](https://en.wikipedia.org/wiki/Interface_segregation_principle). Instead, it could be a generator that yields the required items one by one.

Reply:
no, it works